### PR TITLE
Improve MSP handling

### PR DESCRIPTION
**Pull request by haslinghuis**
- [x] Give FC time to react to new settings
- [x] Reduce MSP.timeout from 250 to 50ms
- [x] Reduce timeout from 250 to 200ms
- [x] Add new mspHelper for EEPROM write

### CODE REVIEW

**Comment by blckmn** [comment]
AUTOMERGE: (FAIL)
 - github identifies PR as mergeable -> FAIL
 - assigned to a milestone -> PASS
 - cooling off period lapsed -> FAIL
 - commit count less or equal to three -> PASS
 - `Don't merge` label NOT found -> PASS
 - at least one `RN:` label found -> PASS
 - `Tested` label found -> FAIL
 - assigned to an approver -> PASS
 - approver count at least three -> FAIL

**Comment by HThuren** [comment]
remove when already are ind queue.... but don't you risk, this case: have a long transaktion, put short status request, another long transaktion, you now want same short status since you still have the answer from first - this are with this PR skipped, and you only get an (old) answer to the short status request.
Better if you refresh the first request, ie, delete and place last in queue, but not if it is a write action ....
Job queues handling are complex to handle...

**Comment by haslinghuis** [comment]
Sure this needs a lot of testing. But see here how it was before: https://github.com/betaflight/betaflight-configurator/commit/e87c0ca4ee3898b2bbbb4504701028f7bbf077bf

**Comment by McGiverGim** [comment]
This code sends the callback? Now we have a MSP request and we inform when finished to the caller. With this PR is discarded, do we inform the caller too?

**Comment by HThuren** [comment]
> This code sends the callback? Now we have a MSP request and we inform when finished to the caller. With this PR is discarded, do we inform the caller too?

PR !, you mean request ? or ...

**Comment by haslinghuis** [comment]
@McGiverGim good call, this was not handled in original code either.

File: src/js/msp.js
Comment (haslinghuis) [reply src/js/msp.js:336]: The code is reverting an experimental fix from @cTn-dev as the concerns are valid but it's how it was https://github.com/betaflight/betaflight-configurator/commit/e87c0ca4ee3898b2bbbb4504701028f7bbf077bf

It's blazing fast - but need to log and test real situations.

:thinking: As alternative for dropping last command - if it exists already - we could replace the existing with the new one.

Comment (haslinghuis) [reply src/js/msp.js]: Reverted the code to apply for MSP V1 requests only.

**Comment by McGiverGim** [comment]
Only as suggestion, now that you are looking at how to accelerate MSP commands, we have a MULTIPLE_MSP command, more info here: https://github.com/betaflight/betaflight-configurator/pull/1605
That can be used to send various commands at once. It is not being used because maybe it can produce some problems with very old FCs if the message is too big, but now that we have abandoned "some" of the backward compatibility, maybe we can try to test it.
Of course, for other PR, not here ;)

**Review by McGiverGim — APPROVED** [review]
Approved, but the MSP_MULTIPLE command that I've mentioned, maybe must be an exception too? It can be used to send different commands to the FC, so it must not be "removed" if there are others in the queue.

**Comment by nerdCopter** [comment]
* `9e265595`
* successfully flashed 2 FC's, both 4.4.1-release and 4.5-zulu
* successfully configured 2 FC's (un-soldered)
* will revisit when @McGiverGim 's concern is resolved.

**Comment by haslinghuis** [comment]
MSP_MULTIPLE_MSP (would call it MSP_BATCH) has yet no examples for implementation.
https://github.com/betaflight/betaflight-configurator/pull/1605

We could look to implement MSP_MULTIPLE_MSP for the status pull (as it's now centralized this would be an easy test).
Firmware buffer size overflow is solved by adding multiple MSP_MULTIPLE_MSP requests so this should certainly not be replaced if not processed before the new request.

:thinking: But as the code already allows an ongoing process to continue adding the second command should already work.

image

**Comment by McGiverGim** [comment]
>MSP_MULTIPLE_MSP (would call it MSP_BATCH) has yet no examples for implementation. #1605

There is one example in the text of the PR. I don't know if it's this what you want.

**Comment by haslinghuis** [comment]
Okay, but we have no production examples / experience - also running parallel (promises) are not recommended according to:

image

**Comment by haslinghuis** [comment]
Using the batched MSP command getting strange behavior:

image

No problem when using the commented code instead.
Also `MSP_SET_CHANNEL_FORWARDING` command does not seem to be used :confused:

**Comment by haslinghuis** [comment]
@McGiverGim for reference (have reverted as it did not work)

image

**Comment by McGiverGim** [comment]
I suggested it for a future PR, not for doing it here, of course ;) but good try.
> also running parallel (promises) are not recommended 

This comment was for the suggestion of doing parallel petitions to the FC that you can see in the thread, but has nothing to do with the MULTIPLE_MSP command (this sends only one petition to the FC containing the data of various commands).

If it does not work, for some reason, maybe something is broken since the time when we add the option. It will need testing. I suggest merge this PR, and after it, try to see if it works and if it is reliable maybe opening a new PR to discuss about it.

**Comment by haslinghuis** [comment]
Only rebased to have the new test units

**Comment by nerdCopter** [comment]
something wrong with save on motors tab

**Comment by haslinghuis** [comment]
@nerdCopter trying to reproduce but works here.

What you are seeing is likely caused by (motors.js):
image

:thinking: without diving in motor data is used on 3 occasions so would require additional refactoring.
For the motor test it's easy, would have to look into motor resource mapping and motor direction tools.
Guess it's outside the scope of this PR but definite would be another performance improvement.

Another occasion would be to have updated motor data would be:
image

**Comment by HThuren** [comment]
Maybe not releated, but my MAG now working, only difference seems to be this PR :-)

**Comment by nerdCopter** [comment]
this was real quad, save dshot300 , then 600.
will re-test unsoldered FC's tomorrow.

https://youtu.be/X_RBpA4_SSo

**Comment by haslinghuis** [comment]
@nerdCopter reduced the timeout letting a duplicate request complete and added debug for which requests are in the queue.
The reduced timeout in rewriting duplicate requests seems to reduce workload in the queue even further.

**Comment by nerdCopter** [comment]
motor-tab not freezing for me anymore.  only slight delay for save & reboot.
few motor-tab tests with console "MSP timeouts" here:
https://youtu.be/zxv1FmXFJIg (tmotor 2-flashes)
https://youtu.be/rfRxCviDOI8 (foxerrf405)

**Comment by haslinghuis** [comment]
@nerdCopter sorted the EEPROM_WRITE - maybe we extend this to other tabs or create a new helper to centralize.

**Comment by nerdCopter** [comment]
nothing is "broken", but i actually have more timeouts this morning. i do not think they are sourced from last commit.  they just exist always but vary depending on the state of the computer's serial subsystem (maybe) -- i think they are sourced from the sensors reading. :man_shrugging: 
image

**Comment by nerdCopter** [comment]
FOUND -- it is indeed the sensors. if i set to 20ms, they timeout warnings completely stop.

**Comment by haslinghuis** [comment]
`MSP_EEPROM_WRITE` stops communication. Converted it to a helper to leave some time to resolve the queue.
Lowered timeout to 200. With value of 100 it became unstable intermittent with loading.

**Comment by nerdCopter** [comment]
`091d7fee`
https://youtu.be/Yq3LI1cSI3Y showing current behavior.  i found when the timeout/queue gets large, moving to another tab lags.

**Comment by haslinghuis** [comment]
We used to have a timeout of 1000ms for MSP timeout as I made it dynamic to be within 250 - 2000ms.
Your video shows it's capable sustaining 200ms as my tests show the same.
Maybe we should limit the refresh rate of sensor polling but I don't think it's a problem for now.

**Comment by nerdCopter** [comment]
if everyone okay with this PR i can approve it.
@McGiverGim ?

**Comment by haslinghuis** [comment]
Needed to refactor ports tab not using async as settings were not applied.

@McGiverGim @nerdCopter not so sure about replacing duplicate MSP requests so dropped that for the moment as clicking other tabs sometimes hung up loading sequence without notice - using a timeout was working but don't want to add another timer as this would still allow duplicates to run.

Speeding up communication works good by improving timing.

**Comment by nerdCopter** [comment]
i have it installed as my daily-driver, will keep an eye on it.

**Review by nerdCopter — APPROVED** [review]
* approving. no known problems.
* MSP timeouts, but as expected.

**Comment by nerdCopter** [comment]
SORRY, converted PR to `draft` to prevent merging.  setting DSHOTXXX does not save

**Review by nerdCopter — APPROVED** [review]
* re-approving
* `c1b1e1a9` fixes last concern.

**Pull request merged**
